feat(EIDOMNI-788): update credential issuance documentation and add m… (#312)
swiyu-admin-ch/swiyu-issuer: migration-guides/guide-2.4.x-3.0.0.md

# Migration Guide: v2.4.x → v3.0.0

This guide helps you migrate from **v2.4.x** to **v3.0.0** of the Generic Issuer Service.

It is based on the **v3.0.0** section in `CHANGELOG.md` ("latest (3.0.0)").

## Compatibility summary

- **Wallets / OID4VCI clients:** may need changes due to removed endpoints and more strictly spec-compliant error
  handling.

> - **Issuer business systems (management API clients / callbacks):** may need changes because some response DTOs and

Reviewer: timing only mentioned here - unclear if it affects us or not.

  callback timing semantics changed.
- **Operators:** should review new environment variables and crypto / metadata validation changes.

## Breaking changes

### 1) `c_nonce` removed from token response (`OAuthTokenDto`)

**What changed**

- `c_nonce` was removed from `OAuthTokenDto`.
- The nonce must now be retrieved from the **nonce endpoint**.
- The `nonce` column was removed from the `credential_offer` table.

**What you need to do**

- If your wallet/client expected `c_nonce` in the token response, update it to call the nonce endpoint.
- Ensure DB migrations have been applied (Flyway) since the schema changed.

### 2) Deprecated OID4VCI Draft 13 endpoints removed

**What changed**

- Draft 13 endpoints are removed.

**What you need to do**

- Update any client integrations still calling draft-13 URLs to the supported OID4VCI 1.0 endpoints (The documentation
  has been updated accordingly).

### 3) `did:jwk` no longer supported

**What changed**

> - Support for `did:jwk` was removed because it is no longer part of the Swiss profiles.

Reviewer 2: ```suggestion
- Support for `did:jwk` for cryptographic_binding_methods_supported in issuer metadata was removed because it is no longer part of the Swiss profiles.
```

**What you need to do**

- Decide whether to enable it and ensure the registry endpoint and credentials are configured.

### 10) Metadata claim descriptor path validation improved

**What changed**

- Validation of metadata claim descriptor paths now correctly supports claims path pointers and validates according to
  specs.
- Metadata claims are for the moment provided under 2 different paths in the metadata :
    - The new `credential_metadata.claims` which:
        - Now uses claims path pointers which are already known from OID4VP DCQL in the verifier
        - If credential metadata claims are set, the issuer validates incomming subject_data against the claim
          descriptors and rejects the request if the validation fails.
    - and the old path `claims` with the same format as before (for backward compatibility reasons).
      **What you need to do**

- If you have custom claim descriptors, validate them against the updated implementation.
- Add the `credential_metadata.claims` to your issuer metadata. Be aware that if you set the claims the validation will
  use these values instead of the legacy `claims` path.

> Old way -> Still used for `claims` as fallback in issuer metadata, but the new way is recommended

Reviewer 2: ```suggestion
Old way -> Still used for `claims` as fallback in issuer metadata, but the new way MUST be used, as the wallet will soon stop support for the old style of claims.
```

            "value_type": "string"
        },
        "average_grade": {
            "mandatory": false,
            "value_type": "number"
        },
        "languages": {
            "mandatory": false,
            "value_type": "array"
        }
    }
}
```

New way with claims path pointers:

```json
{
    "credential_metadata": {
        "claims": [
            {
                "mandatory": true,
                "path": [
                    "type"
                ]
            },
            {
                "mandatory": true,
                "path": [
                    "name"
                ]
            },
            {
                "mandatory": false,
                "path": [
                    "average_grade"
                ]
            },
            {
                "path": [
                    "languages",
                    null
                ]
            }
        ]
    }
}
```

## Upgrade checklist

- [ ] Apply DB migrations (Flyway) and verify the `credential_offer.nonce` column is removed.
- [ ] Update OID4VCI clients to use the nonce endpoint (don’t expect `c_nonce` in token response).
- [ ] Verify no client uses draft-13 endpoints.

> - [ ] Verify issuer DID is not `did:jwk`.

Reviewer 2: ```suggestion
- [ ] Verify issuer metadata cryptographic_binding_methods_supported is not `did:jwk`.
```
